core/src/history/events.rs

```diff
@@ -195,19 +219,54 @@ impl TryFrom<&Path> for EventRegistry {
 }
 
 impl StableXOrderBookReading for EventRegistry {
-    fn get_auction_data<'a>(
-        &'a self,
+    fn get_auction_data(
+        &self,
         batch_id_to_solve: u32,
-    ) -> BoxFuture<'a, Result<(AccountState, Vec<Order>)>> {
-        async move { self.auction_state_for_batch(batch_id_to_solve) }.boxed()
+    ) -> BoxFuture<Result<(AccountState, Vec<Order>)>> {
+        immediate!(self.auction_state_for_batch(batch_id_to_solve))
     }
+
+    fn get_auction_data_for_block(
+        &self,
+        block: BlockNumber,
+    ) -> BoxFuture<Result<(AccountState, Vec<Order>)>> {
+        let (batch_id, block_number) = match block {
+            BlockNumber::Earliest => (BatchId(0), 0),
+            BlockNumber::Latest | BlockNumber::Pending => (BatchId::now(), u64::MAX),
+            // NOTE: Approximate the timestamp of the block by finding the batch
+            // ID of the last event before the specified block.
```

Comment:
Let's put in a doc comment for this method that this doesn't always match what you expect from the documentation on the Trait. If there has been some time without events before this block then the correct batch id might be later than the batch id internally used here.

Reply:
I was under the impression that doc comments don't do anything for trait methods (but I may be wrong).

Reply:
This is unfortunate since it might lead to inconsistencies e.g. when there are future withdraw requests or orders expiring in the future. E.g. if we have an order valid until t and there is no events in t-1, t, t+1. Then if I request the orderbook for a timestamp in t+1 I might still see this order as valid since the batch I'm getting the data for would be t-1 at which point the order was still valid.

Since we are already async here, could we not get the timestamp of this block number to get an exact batchID?